src/engine.rs

```diff
@@ -218,31 +220,38 @@ impl Reedline {
         for command in commands {
             match command {
                 EditCommand::InsertChar(c) => {
-                    let search = self
-                        .history_search
-                        .as_mut()
-                        .expect("couldn't get history_search as mutable"); // We checked it is some
-                    search.step(BasicSearchCommand::InsertChar(*c), &self.history);
+                    let navigation = self.history.get_navigation();
+                    if let HistoryNavigationQuery::SubstringSearch(substring) = navigation {
+                        let new_string = format!("{}{}", substring, c);
+                        self.history
+                            .set_navigation(HistoryNavigationQuery::SubstringSearch(new_string));
+                    } else {
+                        self.history
+                            .set_navigation(HistoryNavigationQuery::SubstringSearch(format!(
+                                "{}",
+                                c
+                            )))
+                    }
                 }
                 EditCommand::Backspace => {
-                    let search = self
-                        .history_search
-                        .as_mut()
-                        .expect("couldn't get history_search as mutable"); // We checked it is some
-                    search.step(BasicSearchCommand::Backspace, &self.history);
-                }
-                EditCommand::SearchHistory => {
-                    let search = self
-                        .history_search
-                        .as_mut()
-                        .expect("couldn't get history_search as mutable"); // We checked it is some
-                    search.step(BasicSearchCommand::Next, &self.history);
+                    let navigation = self.history.get_navigation();
+
+                    if let HistoryNavigationQuery::SubstringSearch(mut substring) = navigation {
+                        let grapheme_left_index = substring[..substring.len()]
+                            .grapheme_indices(true)
+                            .last()
+                            .map(|(i, _)| i)
+                            .unwrap_or(0);
+
+                        substring.replace_range(grapheme_left_index..substring.len(), "");
+
+                        self.history
+                            .set_navigation(HistoryNavigationQuery::SubstringSearch(substring));
+                    }
```

Comment:
Operation like delete single grapheme?

Reply:
I think we have this logic in other places already. Can we re-use the logic the buffer has for the string?

Reply:
I can think of a way to make this into a function and then reuse it here and in line buffer. Thoughts?

Reply:
@sherubthakur sounds good

Reply:
https://github.com/jonathandturner/reedline/pull/92/commits/d36601b0279c2add059a0440f90156ee546bac8d I only extracted the part in engine into a separate function and did not touch line_buffer in this PR. LineBuffer has a lot of custom logic and for at-least now I would like to not worry about unifying it with line_buffer stuff.